field_audit/models.py

```diff
@@ -0,0 +1,259 @@
+from datetime import datetime
+
+from django.conf import settings
+from django.contrib.postgres.indexes import GinIndex
+from django.db import models
+
+from .utils import class_import_helper, get_fqcn
+
+USER_TYPE_TTY = "SystemTtyOwner",
+USER_TYPE_PROCESS = "SystemProcessOwner"
+USER_TYPE_REQUEST = "RequestUser"
+
+
+def check_engine_sqlite(engine=None):
+    """Check if SQLite (or Oracle) database engines are in use. If ``engine`` is
+    ``None``, check all Django engines (otherwise check only ``engine``).
+
+    :param engine: (Optional) name of a Django database engine.
+    """
+    def lite_it_up(engine):
+        # check if engine "flavor" is Oracle or SQLite
+        # example db engine: django.db.backends.sqlite3
+        # resulting flavor:                     sqlite
+        return engine.split(".")[-1][:6] in {"sqlite", "oracle"}
+    if engine is None:
+        for db_properties in settings.DATABASES.values():
+            # The following "no branch" directive prevents coverage from
+            # reporting the (known) untested branch of
+            # `if lite_it_up(...)->return`. There will always be an uncovered
+            # branch here because tests only run on postgres _or_ sqlite, never
+            # both.
+            if lite_it_up(db_properties["ENGINE"]):  # pragma: no branch
+                return True
+    else:
+        return lite_it_up(engine)
+    # "no cover" note: tests only run on postgres _or_ sqlite, never both
+    return False  # pragma: no cover
+
+
+class AuditEventManager(models.Manager):
+    """Manager for the AuditEvent model."""
+
+    def by_type_and_username(self, user_type, username):
+        """Use the ``contains`` query (PostgreSQL and MySQL/MariaDB only) to
+        query for documents with matching keys.
+
+        If other DB flavors are configured in Django settings, support is
+        defined at import time (see below)
+        """
+        # "no cover" note: tests only run on postgres _or_ sqlite, never both
+        return self.filter(  # pragma: no cover
+            changed_by__contains={"user_type": user_type, "username": username},
```

Comment:
this query makes me think that the `changed_by` field expects a certain structure. The docstring for the auditor just says `changed_by() method that returns a JSON-serializable object`. Perhaps that should be constrained to a dataclass if the code is going to rely on specific field in that field.

Reply:
That may be true but then you don't get some of the 'out of box' functionality which seems weird. It should at least be very clearly documented what structure is required in order to get the default queries to work.